torch/csrc/cuda/python_comm.cpp

```diff
@@ -0,0 +1,18 @@
+#include "torch/csrc/utils/pybind.h"
+#include "torch/csrc/cuda/comm.h"
+
+#include <chrono>
+
+namespace torch { namespace cuda { namespace python {
+
+void initCommMethods(PyObject *module) {
+  auto m = py::cast<py::module>(module);
+  m.def("_broadcast_coalesced", [](std::vector<at::Tensor>& tensors, std::vector<int64_t> devices, std::size_t buffer_size) {
```

Comment:
What's the plan with pybind11? Do these methods release the GIL?

FWIW, if you wanted, you could probably bind all these types with `PythonArgParser`, but it doesn't matter.

I think, longer term, it would be good to have the NCCL bindings inside ATen like we do with cuDNN, but it's not necessary in this PR.

Reply:
Oh I forgot they don't (it can be easily fixed by adding an extra argument to `def()`. I'll be happy to use `PythonArgParser` if only you point me to an example

Reply:
There are some examples in the template `torch/csrc/autograd/generated/python_variable_methods.cpp` and more uses in the generated version of that file.

_broadcast_colaesced would looks like:

```c++
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "_broadcast_coalesced(TensorList tensors, IntList devices, int64_t buffer_size=10485760)",
  });
  PyObject* parsed_args[3];
  parser.parse(args, kwargs, parsed_args);
  return wrap(dispatch_broadcast_coalesced(r.tensorlist(0), r.intlist(1), r.toInt64(2)));
  END_HANDLE_TH_ERRORS
```

```c++
static TensorList dispatch_broadcast_coalesced(TensorList tensors, IntList devices, int64_t buffer_size) {
  AutoNoGIL no_gil;
  return broadcast_coalesced(...);
}
```
  
Again, I don't think the choice of binding is particularly important. It's just convenient that it could be written with PythonArgParser because that means that if we push NCCL down to ATen we can generate the Python bindings automatically.

Reply:
That looks really good, I'll use that!

Reply:
This turned out to be more difficult and annoying, especially that we treat all tensors as Variables in this file (which isn't the case here). I'll just leave it in pybind for now.